spec.md

```diff
@@ -1840,6 +1843,49 @@ module MyModule :
    out[4] <= in
 ```
 
+## Bit-indices
+
+The bit-index expression statically refers, by index, to a particular bit, or
+slice of bits, of an expression with an integer type (`UInt`{.firrtl} or
+`SInt`{.firrtl}). The indices must be non-negative integers and cannot be equal
+to or exceed the width of the integer they index. The bit-index `x[hi:lo]`
+selects bits `hi` (most significant) through `lo` (least significant) of `x`.
+The bit-index `x[i]` selects the single bit `i`.
+
+The type of the bit-index expression `x[hi:lo]`.{firrtl} is `UInt<hi - lo +
+1>`.{firrtl} (even if `x` is an `SInt`) and the type of `x[i]`.{firrtl} is
+`UInt<1>`.{firrtl}. This means that when connecting to a bit-indexed value, the
+right-hand-side of the connection must be a `UInt`.{firrtl}, even if the value
+being indexed is an `SInt`.{firrtl}.
+
+The bit-index can be used as a sink or source. When used as a source,
+`x[hi:lo]` is equivalent to `bits(x, hi, lo)` and `x[i]` is equivalent to
+`bits(x, i, i)`. See the `bits`{.firrtl} primitive operation
+([@sec:bit-extraction-operation]). When used as a sink, the bit-index assigns
+to only the sliced bits of the integer. If a value has multiple bit-index
+assignments, the assignments are accumulated in order according to last-connect
+semantics, in the same way as the behavior of last-connect semantics for
+aggregate types (see [@sec:last-connect-semantics]).
+
+A value that is bit-indexed must be fully initialized at the bit-level. There
+must be a valid assignment accounting for every bit in the value. Registers are
+implicitly initialized with their current contents.
+
+Bit-indexing does not participate in width inference (see
+[@sec:width-inference]), and if a bit-index is applied to a value with an
+unspecified width, that value must have another use that allows its width to be
+inferred. Otherwise this causes an error.
```

Comment:
Why not say that a bit-index forces the width to be at least sufficient to the slice bounds?

Reply:
I think that is definitely desirable, but I think the motivation for not participating in width inference was so that the the bit-index would be interchangeable with the `bits` primop (when used as an R-value), and then if/when `bits` gets removed the width inference behavior you describe could be added.

If we want to add width inference support as part of this proposal, then perhaps this proposal should also add width inference for the bits primop.

I think either approach is reasonable.